library/rbtrees.pl

```diff
@@ -141,29 +141,31 @@
 rb_empty(t(Nil,Nil)) :-
     Nil = black('',_,_,'').
 
-%!  rb_lookup(+Key, -Value, +Tree) is semidet.
+%!  rb_lookup(+Key, ?Value, +Tree) is semidet.
 %
 %   True when Value is associated with Key   in the Red-Black tree Tree.
 %   The given Key may include variables, in   which  case the RB tree is
-%   searched for a key with equivalent,   as  in (==)/2, variables. Time
+%   searched for a key with equivalent  variables (using (==)/2).   Time
 %   complexity is O(log N) in the number of elements in the tree.
+%   Throws an error if Key is not sufficiently instantiated (see rb_in/3
+%   if you wish to backtrack over all keys).
 
 rb_lookup(Key, Val, t(_,Tree)) =>
     lookup(Key, Val, Tree).
 
-lookup(_, _, black('',_,_,'')) :- !, fail.
-lookup(Key, Val, Tree) :-
+lookup(Key, _Val, black('',_,_,'')), nonvar(Key) => fail.
+lookup(Key, Val, Tree) =>
```

Comment:
I have my doubt about => in this case as well as lookup/4.   Given a single clause and fully unbound arguments it merely makes the code slower and less traditional for no reason.  For lookup/4 we know the first argument comes from compare/3 and is thus always >, = or <.    IMO the main purpose of => for libraries like this is to get cheap type tests (for the rbtree) and make sure the code is steadfast.

Reply:
The "=>" in lookup/3 is to catch uninstantiated leaves in the tree (which are unlikely; the most common situation is already caught in rb_lookup/3. The original code was a cut/fail, so it's probably about the same level of (in)efficiency as with "=>", unless I misunderstand ssu very badly. 
I've also changed rb_in/3 to use "=>" (same reason as for lookup/3).

As for lookup/4 - I've reverted the "=>"s - they're not necessary by code inspection and they're marginally slower. (Also for next/8 and previous/8)